### ISSUE TITLE
SyncOnSubscribe can call the unsubscription callback multiple times

### ISSUE THREAD

**Issue by akarnokd** [ReactiveX/RxJava#3287]
The following program demonstrates the issue. If a request runs in parallel with a cancellation, both may end up calling the unsubscription callback, which I believe should't happen with the operator.

```java
public class SyncOnSubscribeRace {
    public static void main(String[] args) throws Exception {
        final AtomicInteger wip = new AtomicInteger();
        
        Func0<AtomicInteger> func0 = new Func0<AtomicInteger>() {
            @Override
            public AtomicInteger call() {
                return wip;
            }
        };
        Func2<AtomicInteger, Observer<? super Integer>, AtomicInteger> func2 = 
                new Func2<AtomicInteger, Observer<? super Integer>, AtomicInteger>() {
            @Override
            public AtomicInteger call(AtomicInteger s, Observer<? super Integer> o) { 
                o.onNext(1); 
                return s; 
            }
        };
        Action1<? super AtomicInteger> action1 = new Action1<AtomicInteger>() {
            @Override
            public void call(AtomicInteger s) {
                s.getAndIncrement();
            }
        };
        Observable<Integer> source = Observable.create(
                SyncOnSubscribe.createStateful( 
                func0, 
                func2, action1
        ));

        ExecutorService exec = Executors.newSingleThreadExecutor();
        
        try {
            for (int i = 0; i < 100_000; i++) {
                if (i % 1 == 0) {
                    System.out.println(">> " + i);
                }
                int j = wip.get();
                TestSubscriber<Integer> ts = TestSubscriber.create(0);
                source.subscribe(ts);
                
                final CyclicBarrier cb = new CyclicBarrier(2);
                
                Future<?> f = exec.submit(new Callable<Object>() {
                    @Override
                    public Object call() throws Exception {
                        cb.await();
                        ts.requestMore(1);
                        return null;
                    }
                });
                
                cb.await();
                ts.unsubscribe();
                f.get();
                
                if (j + 1 != wip.get()) {
                    throw new AssertionError("Unsubscribe called multiple times: "
                         + (wip.get() - j));
                }
            }
        } finally {
            exec.shutdownNow();
        }
    }
}
```

**Comment by stealthcode** [comment]
Nice find. Opening a pull request for fix.

**Comment by akarnokd** [comment]
Fixed via #3285.
